From 1e6a8d4f4eaca6efa501fd28f8e385f87d3c7ea8 Mon Sep 17 00:00:00 2001
From: Ilya Ledvich <ilya@compulab.co.il>
Date: Thu, 9 Jun 2016 15:35:07 +0300
Subject: [PATCH 17/38] ARM: dts: imx7d: cl-som-imx7: add analog audio support

Add analog audio support by means of the simple-audio-card.
This instance provides a simple glue interface between cpu:sai and codec:wm8731.
Use a 12.288MHz audio master clock (MCLK) as clock source for an nalog audio
codec wm8731. The MCLK is generated by SAI1 interface and sourced to the
codec from GPIO1_IO01 pin muxed in SAI1_MCLK mode.

Signed-off-by: Ilya Ledvich <ilya@compulab.co.il>
---
 arch/arm/boot/dts/imx7d-cl-som-imx7.dts | 70 ++++++++++++++++++++++++-
 1 file changed, 69 insertions(+), 1 deletion(-)

diff --git a/arch/arm/boot/dts/imx7d-cl-som-imx7.dts b/arch/arm/boot/dts/imx7d-cl-som-imx7.dts
index d64b58a4b6eb..2b5866db921f 100644
--- a/arch/arm/boot/dts/imx7d-cl-som-imx7.dts
+++ b/arch/arm/boot/dts/imx7d-cl-som-imx7.dts
@@ -76,6 +76,35 @@
 			linux,default-trigger = "heartbeat";
 		};
 	};
+
+        sound {
+		compatible = "simple-audio-card";
+		simple-audio-card,name = "cl-som-imx7";
+		simple-audio-card,widgets =
+			"Headphone", "Headphone Jack",
+			"Line", "Line Out",
+			"Microphone", "Mic Jack",
+			"Line", "Line In";
+		simple-audio-card,routing =
+			"Headphone Jack", "RHPOUT",
+			"Headphone Jack", "LHPOUT",
+			"MICIN", "Mic Bias",
+			"Mic Bias", "Mic Jack";
+		simple-audio-card,format = "i2s";
+		simple-audio-card,bitclock-master = <&sound_master>;
+		simple-audio-card,frame-master = <&sound_master>;
+		sound_master: simple-audio-card,cpu {
+			sound-dai = <&sai1 0>;
+			system-clock-direction-out;
+			system-clock-frequency = <0>;
+		};
+
+		simple-audio-card,codec {
+			sound-dai = <&wm8731>;
+			system-clock-frequency = <12288000>;
+		};
+	};
+
 };
 
 &cpu0 {
@@ -316,6 +345,20 @@
 		};
 	};
 
+	wm8731: wm8731@1a {
+		#sound-dai-cells = <0>;
+		compatible = "wlf,wm8731";
+		reg = <0x1a>;
+		clocks = <&clks IMX7D_AUDIO_MCLK_ROOT_CLK>;
+		clock-names = "mclk";
+		assigned-clocks = <&clks IMX7D_AUDIO_MCLK_ROOT_SRC>,
+				  <&clks IMX7D_PLL_AUDIO_POST_DIV>,
+				  <&clks IMX7D_AUDIO_MCLK_ROOT_CLK>;
+		assigned-clock-parents = <&clks IMX7D_PLL_AUDIO_POST_DIV>;
+		assigned-clock-rates = <0>, <884736000>, <12288000>;
+		status = "okay";
+	};
+
 	pca9555: pca9555@20 {
 		compatible = "nxp,pca9555";
 		gpio-controller;
@@ -343,6 +386,17 @@
 	};
 };
 
+&sai1 {
+	#sound-dai-cells = <1>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_sai1 &pinctrl_sai1_lpsr>;
+	assigned-clocks = <&clks IMX7D_SAI1_ROOT_SRC>,
+			  <&clks IMX7D_SAI1_ROOT_CLK>;
+	assigned-clock-parents = <&clks IMX7D_PLL_AUDIO_POST_DIV>;
+	assigned-clock-rates = <0>, <36864000>;
+	status = "okay";
+};
+
 &uart1 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_uart1>;
@@ -506,13 +560,21 @@
 		>;
 	};
 
+	pinctrl_sai1: sai1grp {
+		fsl,pins = <
+			MX7D_PAD_ENET1_RX_CLK__SAI1_TX_BCLK		0x1f
+			MX7D_PAD_ENET1_CRS__SAI1_TX_SYNC		0x1f
+			MX7D_PAD_ENET1_COL__SAI1_TX_DATA0		0x30
+			MX7D_PAD_ENET1_TX_CLK__SAI1_RX_DATA0		0x1f
+		>;
+	};
+
 	pinctrl_tsc2046: tsc2046grp {
 		fsl,pins = <
 			MX7D_PAD_GPIO1_IO14__GPIO1_IO14			0x34 /* PENDOWN */
 		>;
 	};
 
-
 	pinctrl_uart1: uart1grp {
 		fsl,pins = <
 			MX7D_PAD_UART1_TX_DATA__UART1_DCE_TX		0x79
@@ -574,6 +636,12 @@
 		>;
 	};
 
+	pinctrl_sai1_lpsr: sai1grp_lpsr {
+		fsl,pins = <
+			MX7D_PAD_LPSR_GPIO1_IO01__SAI1_MCLK		0x14
+		>;
+	};
+
 	pinctrl_usbotg1: usbotg1grp {
 		fsl,pins = <
 			MX7D_PAD_LPSR_GPIO1_IO05__GPIO1_IO5		0x14 /* OTG PWREN */
-- 
2.20.1

